# Import Dataset (Metro Interstate Traffic Volume)

In [337]:
import pandas as pd
import numpy as np

In [338]:
df = pd.read_csv('dataset/Metro_Interstate_Traffic_Volume.csv')

In [339]:
df.head(len(df))

holiday    temp  rain_1h  snow_1h  clouds_all  weather_main  \
0        None  288.28      0.0      0.0          40        Clouds   
1        None  289.36      0.0      0.0          75        Clouds   
2        None  289.58      0.0      0.0          90        Clouds   
3        None  290.13      0.0      0.0          90        Clouds   
4        None  291.14      0.0      0.0          75        Clouds   
...       ...     ...      ...      ...         ...           ...   
48199    None  283.45      0.0      0.0          75        Clouds   
48200    None  282.76      0.0      0.0          90        Clouds   
48201    None  282.73      0.0      0.0          90  Thunderstorm   
48202    None  282.09      0.0      0.0          90        Clouds   
48203    None  282.12      0.0      0.0          90        Clouds   

          weather_description            date_time  traffic_volume  
0            scattered clouds  2012-10-02 09:00:00            5545  
1               broken clouds  2012-10-02 10:00:00            4516  
2             overcast clouds  2012-10-02 11:00:00            4767  
3             overcast clouds  2012-10-02 12:00:00            5026  
4               broken clouds  2012-10-02 13:00:00            4918  
...                       ...                  ...             ...  
48199           broken clouds  2018-09-30 19:00:00            3543  
48200         overcast clouds  2018-09-30 20:00:00            2781  
48201  proximity thunderstorm  2018-09-30 21:00:00            2159  
48202         overcast clouds  2018-09-30 22:00:00            1450  
48203         overcast clouds  2018-09-30 23:00:00             954  

[48204 rows x 9 columns]

In [340]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   holiday              48204 non-null  object 
 1   temp                 48204 non-null  float64
 2   rain_1h              48204 non-null  float64
 3   snow_1h              48204 non-null  float64
 4   clouds_all           48204 non-null  int64  
 5   weather_main         48204 non-null  object 
 6   weather_description  48204 non-null  object 
 7   date_time            48204 non-null  object 
 8   traffic_volume       48204 non-null  int64  
dtypes: float64(3), int64(2), object(4)
memory usage: 3.3+ MB


In [341]:
df['weather_description'] = df['weather_description'].str.lower()
df['weather_main'] = df['weather_main'].str.lower()

df['date_time'] = df['date_time'].str.slice(0, 10)

df['holiday'] = np.where(df['holiday'] == 'None', "0", "1")

df

holiday    temp  rain_1h  snow_1h  clouds_all  weather_main  \
0           0  288.28      0.0      0.0          40        clouds   
1           0  289.36      0.0      0.0          75        clouds   
2           0  289.58      0.0      0.0          90        clouds   
3           0  290.13      0.0      0.0          90        clouds   
4           0  291.14      0.0      0.0          75        clouds   
...       ...     ...      ...      ...         ...           ...   
48199       0  283.45      0.0      0.0          75        clouds   
48200       0  282.76      0.0      0.0          90        clouds   
48201       0  282.73      0.0      0.0          90  thunderstorm   
48202       0  282.09      0.0      0.0          90        clouds   
48203       0  282.12      0.0      0.0          90        clouds   

          weather_description   date_time  traffic_volume  
0            scattered clouds  2012-10-02            5545  
1               broken clouds  2012-10-02            4516  
2             overcast clouds  2012-10-02            4767  
3             overcast clouds  2012-10-02            5026  
4               broken clouds  2012-10-02            4918  
...                       ...         ...             ...  
48199           broken clouds  2018-09-30            3543  
48200         overcast clouds  2018-09-30            2781  
48201  proximity thunderstorm  2018-09-30            2159  
48202         overcast clouds  2018-09-30            1450  
48203         overcast clouds  2018-09-30             954  

[48204 rows x 9 columns]

In [342]:
holidayDate = list(df[df['holiday'] == 1]['date_time'])

for hd in holidayDate:
    df.loc[df['date_time'] == hd, 'holiday'] = "1"
    
df

holiday    temp  rain_1h  snow_1h  clouds_all  weather_main  \
0           0  288.28      0.0      0.0          40        clouds   
1           0  289.36      0.0      0.0          75        clouds   
2           0  289.58      0.0      0.0          90        clouds   
3           0  290.13      0.0      0.0          90        clouds   
4           0  291.14      0.0      0.0          75        clouds   
...       ...     ...      ...      ...         ...           ...   
48199       0  283.45      0.0      0.0          75        clouds   
48200       0  282.76      0.0      0.0          90        clouds   
48201       0  282.73      0.0      0.0          90  thunderstorm   
48202       0  282.09      0.0      0.0          90        clouds   
48203       0  282.12      0.0      0.0          90        clouds   

          weather_description   date_time  traffic_volume  
0            scattered clouds  2012-10-02            5545  
1               broken clouds  2012-10-02            4516  
2             overcast clouds  2012-10-02            4767  
3             overcast clouds  2012-10-02            5026  
4               broken clouds  2012-10-02            4918  
...                       ...         ...             ...  
48199           broken clouds  2018-09-30            3543  
48200         overcast clouds  2018-09-30            2781  
48201  proximity thunderstorm  2018-09-30            2159  
48202         overcast clouds  2018-09-30            1450  
48203         overcast clouds  2018-09-30             954  

[48204 rows x 9 columns]

In [343]:
df['weather_description'].unique()

array(['scattered clouds', 'broken clouds', 'overcast clouds',
       'sky is clear', 'few clouds', 'light rain',
       'light intensity drizzle', 'mist', 'haze', 'fog',
       'proximity shower rain', 'drizzle', 'moderate rain',
       'heavy intensity rain', 'proximity thunderstorm',
       'thunderstorm with light rain', 'proximity thunderstorm with rain',
       'heavy snow', 'heavy intensity drizzle', 'snow',
       'thunderstorm with heavy rain', 'freezing rain', 'shower snow',
       'light rain and snow', 'light intensity shower rain', 'squalls',
       'thunderstorm with rain', 'proximity thunderstorm with drizzle',
       'thunderstorm', 'very heavy rain',
       'thunderstorm with light drizzle', 'light snow',
       'thunderstorm with drizzle', 'smoke', 'shower drizzle',
       'light shower snow', 'sleet'], dtype=object)

#### Weather_main has it's own categories, and into "weather_description". Firstly, weather_main sort which below,
<br/><br/>
<center> Clear &lt; Clouds &lt; Smoke &lt; Haze &lt; Fog &lt; Mist &lt; Drizzle &lt; Rain &lt; Thunderstorm &lt; Snow &lt; Squall</center>

<br/><br/>
<i> <b>Note: </b> Thuderstorm's categories has to break inside above equation </i>

In [344]:
for wi in list(df['weather_main'].unique()):
    print("=> ", wi)
    for wii in list(df[df['weather_main'] == wi]['weather_description'].unique()):
        print("  --> ", wii)
        
    print("\n")
    print("--------------------------------------------")

=>  clouds
  -->  scattered clouds
  -->  broken clouds
  -->  overcast clouds
  -->  few clouds


--------------------------------------------
=>  clear
  -->  sky is clear


--------------------------------------------
=>  rain
  -->  light rain
  -->  proximity shower rain
  -->  moderate rain
  -->  heavy intensity rain
  -->  freezing rain
  -->  light intensity shower rain
  -->  very heavy rain


--------------------------------------------
=>  drizzle
  -->  light intensity drizzle
  -->  drizzle
  -->  heavy intensity drizzle
  -->  shower drizzle


--------------------------------------------
=>  mist
  -->  mist


--------------------------------------------
=>  haze
  -->  haze


--------------------------------------------
=>  fog
  -->  fog


--------------------------------------------
=>  thunderstorm
  -->  proximity thunderstorm
  -->  thunderstorm with light rain
  -->  proximity thunderstorm with rain
  -->  thunderstorm with heavy rain
  -->  thunderstorm with rain

#### Now, we are using weather_descrition sort which below,

<br><br>

<center>
    clear &lt; few cloud &lt; scattered cloud &lt; broken cloud &lt; overcast cloud &lt; <b> Proximately Thuderstorm </b> &lt; <b> Thuderstorm </b> &lt; smoke &lt; haze &lt; fog &lt; mist &lt; shower drizzle &lt; light intensity drizzle &lt; light drizzle &lt; drizzle &lt; heavy intensity drizzle &lt; heavy drizzle &lt; <b> Proximately Thuderstorm with drizzle &lt; Thuderstorm with light drizzle &lt; Thuderstorm with drizzle &lt; Thuderstorm with heavy drizzle </b> &lt; proximately shower rain &lt; light intensity shower rain &lt; light intensity rain &lt; light rain &lt; moderate rain &lt; heavy intensity rain &lt; very heavy rain &lt; freezing rain &lt; <b> Proximity thuderstorm with rain &lt; thuderstorm with light rain &lt; thuderstorm with rain &lt; thuderstorm with heavy rain </b> &lt; light rain and snow &lt; sleet &lt; light shower snow &lt; shower snow &lt; light snow &lt; snow &lt; heavy snow &lt; squalls
    
</center>
    
<br><br>

<i> <b>Note: </b> Bold text is part of Thuderstorm's categories </i>

In [345]:
lolM = "Clear < Clouds < Smoke < Haze < Fog < Mist < Drizzle < Rain < Thunderstorm < Snow < Squall".lower().split(" < ")

lolM

['clear',
 'clouds',
 'smoke',
 'haze',
 'fog',
 'mist',
 'drizzle',
 'rain',
 'thunderstorm',
 'snow',
 'squall']

In [346]:
lol = "sky is clear < few clouds < scattered clouds < broken clouds < overcast clouds < Proximity Thunderstorm < Thunderstorm < smoke < haze < fog < mist < shower drizzle < light intensity drizzle < light drizzle < drizzle < heavy intensity drizzle < heavy drizzle < Proximity Thunderstorm with drizzle < Thunderstorm with light drizzle < Thunderstorm with drizzle < Thunderstorm with heavy drizzle < Proximity shower rain < light intensity shower rain < light intensity rain < light rain < moderate rain < heavy intensity rain < very heavy rain < freezing rain < Proximity thunderstorm with rain < thunderstorm with light rain < thunderstorm with rain < thunderstorm with heavy rain < light rain and snow < sleet < light shower snow < shower snow < light snow < snow < heavy snow < squalls".lower().split(" < ")

lol

['sky is clear',
 'few clouds',
 'scattered clouds',
 'broken clouds',
 'overcast clouds',
 'proximity thunderstorm',
 'thunderstorm',
 'smoke',
 'haze',
 'fog',
 'mist',
 'shower drizzle',
 'light intensity drizzle',
 'light drizzle',
 'drizzle',
 'heavy intensity drizzle',
 'heavy drizzle',
 'proximity thunderstorm with drizzle',
 'thunderstorm with light drizzle',
 'thunderstorm with drizzle',
 'thunderstorm with heavy drizzle',
 'proximity shower rain',
 'light intensity shower rain',
 'light intensity rain',
 'light rain',
 'moderate rain',
 'heavy intensity rain',
 'very heavy rain',
 'freezing rain',
 'proximity thunderstorm with rain',
 'thunderstorm with light rain',
 'thunderstorm with rain',
 'thunderstorm with heavy rain',
 'light rain and snow',
 'sleet',
 'light shower snow',
 'shower snow',
 'light snow',
 'snow',
 'heavy snow',
 'squalls']

#### Let's convert sort of weather_description into numerical like (0, 1, 2 ---, len(weather_description)-1)

In [347]:
for i in range(0, len(lolM)):
    df['weather_main'].replace(lolM[i], i, inplace=True)

df

holiday    temp  rain_1h  snow_1h  clouds_all  weather_main  \
0           0  288.28      0.0      0.0          40             1   
1           0  289.36      0.0      0.0          75             1   
2           0  289.58      0.0      0.0          90             1   
3           0  290.13      0.0      0.0          90             1   
4           0  291.14      0.0      0.0          75             1   
...       ...     ...      ...      ...         ...           ...   
48199       0  283.45      0.0      0.0          75             1   
48200       0  282.76      0.0      0.0          90             1   
48201       0  282.73      0.0      0.0          90             8   
48202       0  282.09      0.0      0.0          90             1   
48203       0  282.12      0.0      0.0          90             1   

          weather_description   date_time  traffic_volume  
0            scattered clouds  2012-10-02            5545  
1               broken clouds  2012-10-02            4516  
2             overcast clouds  2012-10-02            4767  
3             overcast clouds  2012-10-02            5026  
4               broken clouds  2012-10-02            4918  
...                       ...         ...             ...  
48199           broken clouds  2018-09-30            3543  
48200         overcast clouds  2018-09-30            2781  
48201  proximity thunderstorm  2018-09-30            2159  
48202         overcast clouds  2018-09-30            1450  
48203         overcast clouds  2018-09-30             954  

[48204 rows x 9 columns]

In [348]:
for i in range(0, len(lol)):
    df['weather_description'].replace(lol[i], i, inplace=True)

df

holiday    temp  rain_1h  snow_1h  clouds_all  weather_main  \
0           0  288.28      0.0      0.0          40             1   
1           0  289.36      0.0      0.0          75             1   
2           0  289.58      0.0      0.0          90             1   
3           0  290.13      0.0      0.0          90             1   
4           0  291.14      0.0      0.0          75             1   
...       ...     ...      ...      ...         ...           ...   
48199       0  283.45      0.0      0.0          75             1   
48200       0  282.76      0.0      0.0          90             1   
48201       0  282.73      0.0      0.0          90             8   
48202       0  282.09      0.0      0.0          90             1   
48203       0  282.12      0.0      0.0          90             1   

       weather_description   date_time  traffic_volume  
0                        2  2012-10-02            5545  
1                        3  2012-10-02            4516  
2                        4  2012-10-02            4767  
3                        4  2012-10-02            5026  
4                        3  2012-10-02            4918  
...                    ...         ...             ...  
48199                    3  2018-09-30            3543  
48200                    4  2018-09-30            2781  
48201                    5  2018-09-30            2159  
48202                    4  2018-09-30            1450  
48203                    4  2018-09-30             954  

[48204 rows x 9 columns]

In [349]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   holiday              48204 non-null  object 
 1   temp                 48204 non-null  float64
 2   rain_1h              48204 non-null  float64
 3   snow_1h              48204 non-null  float64
 4   clouds_all           48204 non-null  int64  
 5   weather_main         48204 non-null  int64  
 6   weather_description  48204 non-null  int64  
 7   date_time            48204 non-null  object 
 8   traffic_volume       48204 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 3.3+ MB


# Feature extraction using Correlation

## For Numerical

In [350]:
from scipy.stats import pearsonr

In [351]:
numLst = ["temp", "rain_1h", "snow_1h", "clouds_all"]
tar = "traffic_volume"

resLst = []

In [352]:
for nl in numLst:
    corr, _ = pearsonr(df[tar], df[nl])
    
    if corr > 0.5 or corr < -0.5:
        resLst.append(
            [nl, corr, "Success"]
        )
        
    else:
        resLst.append(
            [nl, corr, "Fail"]
        )

In [353]:
resLst = pd.DataFrame(resLst)
resLst.columns = ["Attribute", "Correlation", "Status"]

resLst

Attribute  Correlation Status
0        temp     0.130299   Fail
1     rain_1h     0.004714   Fail
2     snow_1h     0.000733   Fail
3  clouds_all     0.067054   Fail

## For Categorical

In [354]:
from scipy import stats

In [355]:
catLst = ['holiday', 'weather_main', 'weather_description', 'date_time']

resLst = []

In [356]:
for lol in catLst:
    res = [lol]
    
    groupedTest = df[[lol, tar]].groupby([lol])
    unique_ = df[lol].unique()
    
    f_val, p_val = stats.f_oneway(
        *(groupedTest.get_group(value)[tar] for value in unique_)
    )
    
    res.append(f_val)
    res.append(p_val)
    
    dfn = len(groupedTest) - 1
    dfd = 0
    for value in unique_:
        dfd += groupedTest.get_group(value)[tar].shape[0]
        
    res.append(
        stats.f.ppf(q=1-0.05, dfn=dfn, dfd=dfd)
    )
    
    resLst.append(res)

In [357]:
resLst = pd.DataFrame(resLst)
resLst.columns = ["Attribute", "F-test", "P-value", "Results"]

resLst

Attribute     F-test        P-value   Results
0              holiday  88.862891   4.412191e-21  3.841652
1         weather_main  97.411612  8.353544e-201  1.830900
2  weather_description  34.133716  1.566607e-231  1.416874
3            date_time   2.675259  7.075537e-261  1.055678

### Do Data Transformation for categorical variables

In [358]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   holiday              48204 non-null  object 
 1   temp                 48204 non-null  float64
 2   rain_1h              48204 non-null  float64
 3   snow_1h              48204 non-null  float64
 4   clouds_all           48204 non-null  int64  
 5   weather_main         48204 non-null  int64  
 6   weather_description  48204 non-null  int64  
 7   date_time            48204 non-null  object 
 8   traffic_volume       48204 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 3.3+ MB


In [360]:
df['holiday'] = df['holiday'].astype("int")
df['weather_description'] = df['weather_description'].astype("int")
df['weather_main'] = df['weather_main'].astype("int")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   holiday              48204 non-null  int64  
 1   temp                 48204 non-null  float64
 2   rain_1h              48204 non-null  float64
 3   snow_1h              48204 non-null  float64
 4   clouds_all           48204 non-null  int64  
 5   weather_main         48204 non-null  int64  
 6   weather_description  48204 non-null  int64  
 7   date_time            48204 non-null  object 
 8   traffic_volume       48204 non-null  int64  
dtypes: float64(3), int64(5), object(1)
memory usage: 3.3+ MB


# Modeling

## Simple Linear Regression

In [361]:
from sklearn.linear_model import LinearRegression

lm = LinearRegression()
x = df[['temp']]
y = df[tar]

lm.fit(x, y)

LinearRegression()

In [362]:
print("Intercept: ", lm.intercept_, " whereas, slope: ", lm.coef_)

Intercept:  -2198.186797064827  whereas, slope:  [19.40928597]


Thus, final extimated model in traffic_volume = -2198.1868 + (19.4093) * temp

In [363]:
yHat = lm.predict(x) 

print("Predicated Value\tActual Value\n")
for i in range(0, 5):
    print(yHat[i], "\t", y[i])

Predicated Value	Actual Value

3397.122161617313 	 5545
3418.084190462106 	 4516
3422.3542333749333 	 4767
3433.029340657004 	 5026
3452.6327194840783 	 4918


## Multiple Linear Regression

In [364]:
zList = []

for b in list(df.columns):
    if b != tar and b != "date_time":
        zList.append(b)
        
z = df[[
    *(x for x in zList)
]]

In [365]:
lm = LinearRegression()

lm.fit(z, df[tar])

LinearRegression()

In [366]:
print("Intercept: ", lm.intercept_, " whereas, slope: ", lm.coef_)

Intercept:  -2999.222245113267  whereas, slope:  [-2.30586309e+03  2.17945193e+01  1.60819004e-01  7.75371861e+02
  7.07651495e+00 -1.23954347e+02  1.64354761e+01]


In [367]:
yHatMulti = lm.predict(z)

print("Predicated value\tActual value\n")

for i in range(0, 5):
    print(yHatMulti[i], "\t", y[i])

Predicated value	Actual value

3475.6789792574696 	 5545
3763.330559352096 	 4516
3890.7085539158766 	 4767
3902.6955395269547 	 5026
3802.1248036934016 	 4918


## Polynomial Regression

In [368]:
x = df['temp']
y = df[tar]

f = np.polyfit(x, y, 3)
p = np.poly1d(f)

print(p)

          3         2
0.002493 x - 1.345 x + 187.8 x + 1200


In [369]:
yPolyPred = p(x)

print("Predicated\tActual\n")
for i in range(0, 5):
    print(yPolyPred[i], "\t", y[i])

Predicated	Actual

3340.5209895807016 	 5545
3378.440570403391 	 4516
3386.398796947804 	 4767
3406.6426205950115 	 5026
3445.120300197763 	 4918


## Multiple Polynomial Regression

In [370]:
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.pipeline import Pipeline

In [371]:
ipt = [('Scale', StandardScaler()), ('Polynomial', PolynomialFeatures(2)), ('Model', LinearRegression())]

pipe = Pipeline(ipt)
pipe

Pipeline(steps=[('Scale', StandardScaler()),
                ('Polynomial', PolynomialFeatures()),
                ('Model', LinearRegression())])

In [372]:
pipe.fit(z, y)

Pipeline(steps=[('Scale', StandardScaler()),
                ('Polynomial', PolynomialFeatures()),
                ('Model', LinearRegression())])

In [373]:
yPipe = pipe.predict(z)

print("Predicated\tActual\n")
for i in range(0, 5):
    print(yPipe[i], "\t", y[i])

Predicated	Actual

3836.802764892578 	 5545
3930.2606201171875 	 4516
3760.955612182617 	 4767
3777.313461303711 	 5026
3981.920944213867 	 4918


# MSE and R^2

In [374]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [376]:
print("Simple Linear Regression")
print("R^2: ", r2_score(df[tar], yHat))
print("MSE: ", mean_squared_error(df[tar], yHat))

print("-----------------------------------------------------------------")

print("Multiple Linear Regression")
print("R^2: ", r2_score(df[tar], yHatMulti))
print("MSE: ", mean_squared_error(df[tar], yHatMulti))

print("-----------------------------------------------------------------")

print("Simple Polynomail Regression")
print("R^2: ", r2_score(df[tar], yPolyPred))
print("MSE: ", mean_squared_error(df[tar], yPolyPred))

print("-----------------------------------------------------------------")

print("Mulitple Polynomial Regression")
print("R^2: ", r2_score(df[tar], yPipe))
print("MSE: ", mean_squared_error(df[tar], yPipe))

Simple Linear Regression
R^2:  0.01697777680484025
MSE:  3880513.0883285766
-----------------------------------------------------------------
Multiple Linear Regression
R^2:  0.03477460746877037
MSE:  3810259.504337691
-----------------------------------------------------------------
Simple Polynomail Regression
R^2:  0.021908921826624073
MSE:  3861047.2285077455
-----------------------------------------------------------------
Mulitple Polynomial Regression
R^2:  0.046663375208083946
MSE:  3763328.1962472675
